server/src/test/java/org/elasticsearch/cluster/coordination/PublicationTransportHandlerTests.java

```diff
@@ -330,4 +338,160 @@ public void writeTo(StreamOutput out) throws IOException {
+                new ClusterStatePublicationEvent(
+                    new BatchSummary("test"),
+                    clusterState0,
+                    clusterState0,
+                    new Task(randomNonNegativeLong(), "test", "test", "", TaskId.EMPTY_TASK_ID, Map.of()),
+                    0L,
+                    0L
+                )
+            );
+        try {
+            context0.sendPublishRequest(
+                otherNode,
+                new PublishRequest(clusterState0),
+                ActionListener.wrap(() -> assertTrue(completed.compareAndSet(false, true)))
+            );
+            assertTrue(completed.getAndSet(false));
+            var receivedState = receivedStateRef.getAndSet(null);
+            assertEquals(clusterState0.stateUUID(), receivedState.stateUUID());
+            assertEquals(otherNode, receivedState.nodes().getLocalNode());
+            assertFalse(receivedState.metadata().clusterUUIDCommitted());
+            assertEquals(VotingConfiguration.of(), receivedState.getLastCommittedConfiguration());
+        } finally {
+            context0.decRef();
+        }
+
+        final var committedClusterState0 = ClusterState.builder(clusterState0)
+            .metadata(clusterState0.metadata().withLastCommittedValues(true, clusterState0.getLastAcceptedConfiguration()))
+            .build();
+        assertEquals(clusterState0.stateUUID(), committedClusterState0.stateUUID());
+        assertEquals(clusterState0.term(), committedClusterState0.term());
+        assertEquals(clusterState0.version(), committedClusterState0.version());
+
+        final var clusterState1 = ClusterState.builder(committedClusterState0).incrementVersion().build();
+        assertSame(committedClusterState0.metadata(), clusterState1.metadata());
+
+        var context1 = transportHandlersByNode.get(localNode)
+            .newPublicationContext(
+                new ClusterStatePublicationEvent(
+                    new BatchSummary("test"),
+                    committedClusterState0,
+                    clusterState1,
+                    new Task(randomNonNegativeLong(), "test", "test", "", TaskId.EMPTY_TASK_ID, Map.of()),
+                    0L,
+                    0L
+                )
+            );
+        try {
+            context1.sendPublishRequest(
+                otherNode,
+                new PublishRequest(clusterState1),
+                ActionListener.wrap(() -> assertTrue(completed.compareAndSet(false, true)))
+            );
+            assertTrue(completed.getAndSet(false));
+            var receivedState = receivedStateRef.getAndSet(null);
+            assertEquals(clusterState1.stateUUID(), receivedState.stateUUID());
+            assertEquals(otherNode, receivedState.nodes().getLocalNode());
+            assertTrue(receivedState.metadata().clusterUUIDCommitted());
+            assertEquals(VotingConfiguration.of(localNode), receivedState.getLastCommittedConfiguration());
```

Comment:
Can we add something like:

```
assertSame(receivedState.nodes(), firstReceivedState.nodes());
```
where `firstReceivedState` is the `receivedState` from the `context0` invocation above? It is mostly a test validation, i.e., that we do send a diff in this case (which AFAICS, we will).

Reply:
Yep, and we can see that from the receiving node's `PublishClusterStateStats` too. See b910c87aaed.